stdlib/mexpr/boot-parser.mc

```diff
@@ -0,0 +1,258 @@
+  sem matchTerm (t:Unknown) =
+  | 100 /-TmVar-/ ->
+      TmVar {ident = gname t 0,
+             ty = TyUnknown(),
+             info = ginfo t}
+  | 101 /-TmLam-/ ->
+      TmLam {ident = gname t 0,
+             ty = gtype t 0,
+             info = ginfo t,
+             body = gterm t 0}
+  | 102 /-TmLet-/ ->
+      TmLet {ident = gname t 0,
+             tyBody = gtype t 0,
+             body = gterm t 0,
+             inexpr = gterm t 1,
+             ty = TyUnknown(),
+             fi = ginfo t}
+  | 103 /-TmType-/ ->
+      TmType {ident = gname t 0,
+              ty = gtype t 0,
+              inexpr = gterm t 0,
+              fi = ginfo t}
+  | 104 /-TmRecLets-/ ->
+      TmRecLets {bindings =
+                   makeSeq (lam n. {ident = gname t n,
+                                 ty = gtype t n,
+                                 body = gterm t n}) (glistlen t 0),
+                 inexpr = gterm t (glistlen t 0),
+                 ty = TyUnknown()}                            
+  | 105 /-TmApp-/ ->
+      TmApp {lhs = gterm t 0,
+             rhs = gterm t 1,
+             ty = TyUnknown(),
+             fi = ginfo t}     
+  | 106 /-TmConst-/ ->
+      let c = gconst t 0 in
+      TmConst {val = gconst t 0,
+               ty = match c with CBool _ then TyBool {} else
+                    match c with CInt _ then TyInt {} else
+                    match c with CFloat _ then TyFloat {} else
+                    TyChar {},
+               fi = ginfo t}
+  | 107 /-TmSeq-/ ->
+      TmSeq {tms = makeSeq (lam n. gterm t n) (glistlen t 0),
+             ty =  TyUnknown(),
+             fi = ginfo t}
+  | 108 /-TmRecord-/ ->
+     let lst = makeSeq (lam n. (gstr t n, gterm t n)) (glistlen t 0) in
+      TmRecord {bindings = seq2assoc {eq = eqString} lst,
+               ty = TyUnknown(),
+               fi = ginfo t}
+  | 109 /-TmRecordUpdate-/ ->
+     TmRecordUpdate {rec = gterm t 0,
+                    key = gstr t 0,
+                    value = gterm t 1,
+                    ty = TyUnknown(),
+                    fi = ginfo t}
+  | 110 /-TmConDef-/ ->
+     TmConDef {ident = gname t 0,
+               ty = gtype t 0,
+               inexpr = gterm t 0}
+  | 111 /-TmConApp-/ ->
+     TmConApp {ident = gname t 0,
+               body = gterm t 0,
+               ty = TyUnknown()}
+  | 112 /-TmMatch-/ ->
+     TmMatch {target = gterm t 0,
+              pat = gpat t 0,
+              thn = gterm t 1,
+              els = gterm t 2,
+              ty = TyUnknown(),
+              fi = ginfo t}
+  
+  sem gpat (t:Unkown) =
+  | n -> let t2 = bootParserGetPat t n in
+         matchPat t2 (bootParserGetId t2)
+
+  sem matchPat (t:Unknown) =
+  | 400 /-PatNamed-/ -> strToPatName (gstr t 0) 
+  | 401 /-PatSeqTot-/ ->
+    PSeqTot {pats = makeSeq (lam n. gpat t n) (glistlen t 0) }
+  | 402 /-PatSeqEdge-/ ->
+    let len = glistlen t 0 in
+    PSeqEdge {prefix = makeSeq (lam n. gpat t n) len,
+              middle = strToPatName (gstr t 0),
```

Comment:
Related to the other comment on `strToPatName`, this should be a `PatName`, not a `Pat`.